src/thumbnailer_slope_optim.cc

```diff
@@ -0,0 +1,253 @@
+            [](const FrameData& a, const FrameData& b) -> bool {
+              return a.timestamp_ms < b.timestamp_ms;
+            });
+
+  float limit_slope;
+  if (FindMedianSlope(&limit_slope) != kOk) {
+    return kSlopeOptimError;
+  }
+
+  int min_quality = minimum_lossy_quality_;
+  int max_quality = 100;
+  WebPData new_webp_data;
+  WebPDataInit(&new_webp_data);
+
+  std::queue<int> optim_list;  // Queue of frames needed to find the quality in
+                               // the next binary search loop.
+  for (int i = 0; i < frames_.size(); i++) {
+    optim_list.push(i);
+  }
+
+  // Use binary search with slope optimization to find quality values that makes
+  // the animation fit the given byte budget. The quality value for each frame
+  // can be different.
+  while (min_quality <= max_quality && !optim_list.empty()) {
+    int mid_quality = (min_quality + max_quality) / 2;
+
+    std::queue<int> new_optim_list;
+
+    while (!optim_list.empty()) {
+      int curr_frame = optim_list.front();
+      optim_list.pop();
+
+      float curr_slope;
+      if (ComputeSlope(curr_frame, min_quality, max_quality, &curr_slope) !=
+          kOk) {
+        return kSlopeOptimError;
+      }
+
+      if (frames_[curr_frame].final_quality == -1 || curr_slope > limit_slope) {
+        frames_[curr_frame].config.quality = mid_quality;
+        new_optim_list.push(curr_frame);
+      }
+    }
+
+    if (new_optim_list.empty()) break;
+
+    if (GenerateAnimationNoBudget(&new_webp_data) != kOk) {
+      return kMemoryError;
+    }
+
+    if (new_webp_data.size <= byte_budget_) {
+      while (!new_optim_list.empty()) {
+        int curr_frame = new_optim_list.front();
+        new_optim_list.pop();
+        frames_[curr_frame].final_quality = mid_quality;
+        optim_list.push(curr_frame);
+      }
+      WebPDataClear(webp_data);
+      *webp_data = new_webp_data;
+      min_quality = mid_quality + 1;
+    } else {
+      optim_list = new_optim_list;
+      max_quality = mid_quality - 1;
+      WebPDataClear(&new_webp_data);
+    }
+  }
+
+  std::cout << "Final qualities with slope optimization: ";
+
+  for (auto& frame : frames_) {
+    frame.config.quality = frame.final_quality;
+    if (GetPictureStats(frame.pic, frame.config, &frame.encoded_size,
+                        &frame.final_psnr) != kOk) {
+      return kStatsError;
+    }
+    std::cerr << frame.config.quality << " ";
+  }
+
+  std::cout << std::endl;
+
+  return (webp_data->size > 0) ? kOk : kByteBudgetError;
+}
+
+Thumbnailer::Status Thumbnailer::LossyEncodeNoSlopeOptim(
+    WebPData* const webp_data) {
+  int anim_size = 0;
+  for (auto& frame : frames_) {
+    anim_size += frame.encoded_size;
+  }
+
+  WebPData new_webp_data;
+  WebPDataInit(&new_webp_data);
+
+  // For each frame, find the best quality value that can produce the higher
+  // PSNR than the current one if possible.
+  for (auto& frame : frames_) {
+    int min_quality = 0;
+    int max_quality = 100;
```

Comment:
Is it really necessary to search over the whole range? Maybe it's enough to search around a window of, say, +/- 10 around the existing quality setting?